File: python/src/task_percentage.py
from fractions import Fraction
common_divisors = [9, 7, 5, 3, 2]
absolute_total = 100

def to_seconds(hour_format: str):
	hour, minutes, seconds = hour_format.split(":")

	min_to_sec = int(minutes) * 60
	hr_to_sec = int(hour) * 60 * 60

	total_seconds = int(seconds) + min_to_sec + hr_to_sec
	return total_seconds

def min_common_divisor(num1: int, num2: int):
	for divisor in common_divisors:
		if num1 % divisor == 0 and num2 % divisor == 0:
			return divisor

def task_percentage(partial: str, total: str):
	partial = to_seconds(partial)
	total = to_seconds(total)

	partial_percentage = (partial * absolute_total) / total
	rounded_percentage = round(absolute_total / partial_percentage, 1)

	if (rounded_percentage % 1 != 0):
		current_total = min_common_divisor(partial_percentage, absolute_total)
		current_partial = (partial_percentage * current_total) / absolute_total
		return f"{round(current_partial)}/{current_total}"
	else:
		return f"1/{round(rounded_percentage)}"


if __name__ == "__main__":
	print(
		task_percentage('01:00:00', '03:00:00')
	) # '1/3'
	print(
		task_percentage('02:00:00', '04:00:00')
	) # '1/2'
	print(
		task_percentage('01:00:00', '01:00:00')
	) # '1/1'
	print(
		task_percentage('00:10:00', '01:00:00')
	) # '1/6'
	print(
		task_percentage('01:10:10', '03:30:30')
	) # '1/3'
	print(
		task_percentage('03:30:30', '05:50:50')
	) # '3/5

File: typescript/src/taskPercentage.ts
const commonDivisors = [9, 7, 5, 3, 2]
const absoluteTotal = 100

function toSeconds(hourFormat: string) {
	const times = hourFormat.split(":").map(time => Number(time))
	const hour = times[0]
	const minutes = times[1]
	const seconds = times[2]

	const minToSec = minutes * 60
	const hrToSec = hour * 60 * 60

	const totalSeconds = seconds + minToSec + hrToSec
	return totalSeconds
}

function minCommonDivisor(num1: number, num2: number) {
	for (let divisor of commonDivisors) {
		if (num1 % divisor == 0 && num2 % divisor == 0) {
			return divisor
		}
	}

	return 0
}

function taskPercentage(partial: string | number, total: string | number) {
	partial = toSeconds(partial as string)
	total = toSeconds(total as string)

	const partialPercentage = (partial * absoluteTotal) / total
	const roundedPercentage = Number(Math.round((absoluteTotal / partialPercentage) * 10) / 10)

	if (roundedPercentage % 1 != 0) {
		const currentTotal = minCommonDivisor(partialPercentage, absoluteTotal)
		const currentPartial = (partialPercentage * currentTotal) / absoluteTotal
		return `${Math.round(currentPartial)}/${currentTotal}`
	} else {
		return `1/${Math.round(roundedPercentage)}`
	}
}

console.log(
	taskPercentage('01:00:00', '03:00:00') // '1/3'
)
console.log(
	taskPercentage('02:00:00', '04:00:00') // '1/2'
)
console.log(
	taskPercentage('01:00:00', '01:00:00') // '1/1'
)
console.log(
	taskPercentage('00:10:00', '01:00:00') // '1/6'
)
console.log(
	taskPercentage('01:10:10', '03:30:30') // '1/3'
)
console.log(
	taskPercentage('03:30:30', '05:50:50') // '3/5
)